Fix issue DPTP-3299 Add HyperShift installation files (#37500)
openshift/release: clusters/hive/hypershift/README.md

This directory includes the files (include generated files) for deploying HyperShift on `hive`.

# HyperShift CLI and operator images
HyperShift CLI (the utility for installing HyperShift, as well as creting/destroying hosted clusters) and HyperShift operator (the core component on HyperShift management cluster) share the same image, due to the [fact](https://issues.redhat.com/browse/HOSTEDCP-792) that HyperShift does not provide backward or forward compatibilities.

The image is built with `clusters/app.ci/supplemental-ci-images/hypershift/hypershift-cli.yaml` to available for all build farms.

# Deployment/upgrading on `hive`
Generate the deployment file with
```

> docker run -ti --rm -v ~/workspace/mgmt-aws:/mgmt-aws:ro quay.io/hypershift/hypershift-operator:latest install --oidc-storage-provider-s3-bucket-name=hypershift-oidc-provider --oidc-storage-provider-s3-credentials=/mgmt-aws --oidc-storage-provider-s3-region=us-east-1 --hypershift-image=https://registry.ci.openshift.org/ci/hypershift-cli:latest --enable-uwm-telemetry-remote-write=false render

Reviewer: Does this work out of the box? It seems you're mapping from `~/workspace/mgmt-aws` but if have none locally

Author: Added a line to clarify that.

Reviewer: Thanks

This directory includes the files (include generated files) for deploying HyperShift on `hive`.

# HyperShift CLI and operator images
HyperShift CLI (the utility for installing HyperShift, as well as creting/destroying hosted clusters) and HyperShift operator (the core component on HyperShift management cluster) share the same image, due to the [fact](https://issues.redhat.com/browse/HOSTEDCP-792) that HyperShift does not provide backward or forward compatibilities.

> The image is built with `clusters/app.ci/supplemental-ci-images/hypershift/hypershift-cli.yaml` to available for all build farms.

Reviewer 2: - Why do we need to build it? As we need `jq`?
- Do we get the operator from the same image? How can we ensure it is the same version of cli, e.g., someone triggered a build from the BC?
- Could you explain how we ensure that they have the same version on our production? Does the workflow use the cli?

> The image is built with `clusters/app.ci/supplemental-ci-images/hypershift/hypershift-cli.yaml` to available for all build farms.

Author: > Why do we need to build it? As we need jq?
we need `jq` and we need a stable version. The image on `quay.io` are not tagged and untagged versions are quickly purged.

> Do we get the operator from the same image? How can we ensure it is the same version of cli, e.g., someone triggered a build from the BC?
No one, unless the want to upgrade HyperShift versions, should trigger builds for this BC. This is special.

> Could you explain how we ensure that they have the same version on our production? Does the workflow use the cli?
The workflow is using this CLI as well.

> The image is built with `clusters/app.ci/supplemental-ci-images/hypershift/hypershift-cli.yaml` to available for all build farms.

Reviewer 2: > No one, unless they want to upgrade HyperShift versions, should trigger builds for this BC. This is special.

Let us create a new tag to reserve the version we use in our production.

> The workflow is using this CLI as well.

Please clarify it in the doc.

> The image is built with `clusters/app.ci/supplemental-ci-images/hypershift/hypershift-cli.yaml` to available for all build farms.

Reviewer 2: > we need jq and we need a stable version. The image on quay.io are not tagged and untagged versions are quickly purged.

Please clarify this in the doc.

# Deployment/upgrading on `hive`

> Prepare the AWS credentials for your management cluster account in `~/workspace/mgmt-aws` (or any other path you prefer).

Reviewer 2: - Did you use personal credentials? I'd like to know what to do if I need to redo hypershift.
- Could we run the cmd without `~/workspace/mgmt-aws`? Easy for copy/paste.

Author: > Did you use personal credentials? I'd like to know what to do if I need to redo hypershift.
the `mgmt-aws` credential is the credential for the AWS account for the management cluster. In this case, its a credential to the AWS account hosting `hive`. Its one time used only (during the installation).

> Could we run the cmd without ~/workspace/mgmt-aws? Easy for copy/paste.
yes, but we still need to specify the files...

Reviewer 2: The credentials to which AM user? Under which AWS account?
Let us clarify it in the doc if it does not expose sensitive information.

>yes, but we still need to specify the files...

I do not understand this.

> Generate the deployment file with

Reviewer 2: ```suggestion
Generate the Hypershift deployment manifest `hypershift.yaml`:
```

```
docker run -ti --rm -v ~/workspace/mgmt-aws:/mgmt-aws:ro registry.ci.openshift.org/ci/hypershift-cli:latest install --oidc-storage-provider-s3-bucket-name=hypershift-oidc-provider --oidc-storage-provider-s3-credentials=/mgmt-aws --oidc-storage-provider-s3-region=us-east-1 --hypershift-image=https://registry.ci.openshift.org/ci/hypershift-cli:latest --enable-uwm-telemetry-remote-write=false render
```

Then apply the file with
`oc --context hive apply -f hypershift.yaml`

# Note
!!! IMPORTANT !!!

> The CLI used to generate the deployment file must be the _EXACT_ CLI used to create/destroy hosted cluster.

Reviewer 2: - Could you add the command to get the cli?

Author: I believe thats already included
```
docker run -ti --rm -v ~/workspace/mgmt-aws:/mgmt-aws:ro registry.ci.openshift.org/ci/hypershift-cli:latest install --oidc-storage-provider-s3-bucket-name=hypershift-oidc-provider --oidc-storage-provider-s3-credentials=/mgmt-aws --oidc-storage-provider-s3-region=us-east-1 --hypershift-image=https://registry.ci.openshift.org/ci/hypershift-cli:latest --enable-uwm-telemetry-remote-write=false render
```

Reviewer 2: The `cli` here is docker, no?

This directory includes the manifests, including the generated ones, for deploying HyperShift on the `hive` cluster.

# HyperShift CLI and operator images
[HyperShift CLI](https://hypershift-docs.netlify.app/getting-started/) which is the utility for installing HyperShift as well as creating and destroying hosted clusters and the HyperShift operator which is the core component running on HyperShift management cluster share the same image. 
They must have the exact same version to work properly [HOSTEDCP-792](https://issues.redhat.com/browse/HOSTEDCP-792).

> The image is built with [hypershift-cli.yaml](clusters/app.ci/supplemental-ci-images/hypershift/hypershift-cli.yaml) to available for all build farms.

Reviewer 2: ```suggestion
The image is built with [hypershift-cli.yaml](clusters/app.ci/supplemental-ci-images/hypershift/hypershift-cli.yaml). The purposes of the build are
- add useful utilities, e.g., `jq`, and
- save a stable tag as explained before.
```

# Deployment/upgrading on `hive`

> Prepare the AWS credentials for your management cluster account (in this case, the AWS account running `hive` cluster) in `~/workspace/mgmt-aws` (or any other path you prefer).

Reviewer 2: ```suggestion
Prepare the AWS credentials for the management cluster account (in this case, the AWS account running `hive` cluster) in `~/workspace/mgmt-aws` (or any other path you prefer).
```

Reviewer 2: My comment https://github.com/openshift/release/pull/37500#discussion_r1164508883 is still not addressed. :)

Generate the Hypershift deployment manifest `hypershift.yaml`:
```
docker run -ti --rm -v ~/workspace/mgmt-aws:/mgmt-aws:ro registry.ci.openshift.org/ci/hypershift-cli:<tag> install --oidc-storage-provider-s3-bucket-name=hypershift-oidc-provider --oidc-storage-provider-s3-credentials=/mgmt-aws --oidc-storage-provider-s3-region=us-east-1 --hypershift-image=https://registry.ci.openshift.org/ci/hypershift-cli:<tag> --enable-uwm-telemetry-remote-write=false render > hypershift-install.yaml
```

The image in parameter `--hypershift-image` must have the SAME as the one invoking the command.

Then apply the file with

> `oc --context hive apply -f hypershift.yaml --server-side`

Reviewer 2: ```suggestion
`oc --context hive apply -f hypershift-install.yaml --server-side`
```

> Dont forget to update the workflow (`ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml`) to use the new image.

Reviewer 2: ```suggestion
The image in the workflow (`ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml`) has to be modifed accordingly each time after a new version of `hypershift-install.yaml` is generated and applied..
```
